File: interceptor/tests/test_interceptor.mjs
/**
 * JS 拦截器的行为测试（Node 跑，不需要浏览器）。
 *
 *   node tests/test_interceptor.mjs
 *
 * 重点测三件事:
 *   1. 该改的改 (api.hypixel.net -> 反代), 不该改的一个都别动
 *   2. key 的注入 / 强制覆盖 / 不覆盖 三种策略
 *   3. fetch 的两种调用形态 (字符串 / Request 对象) 都真的被拦到
 */
import assert from 'node:assert';
import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);
const BSK = require('../bsk-hypixel-interceptor.js');

let pass = 0, fail = 0;
async function t(name, fn) {
  try {
    await fn();
    pass++;
    console.log('  ✓ ' + name);
  } catch (e) {
    fail++;
    console.log('  ✗ ' + name + '\n      ' + e.message);
  }
}

const KEY = 'bsk_TESTKEY';

console.log('URL 改写');
BSK.setConfig({ proxyBase: 'https://hyp-api.firebounce.today', apiKey: KEY, forceKey: true });

await t('官方域名被改到反代', () => {
  assert.strictEqual(
    BSK.rewriteUrl('https://api.hypixel.net/v2/player?uuid=abc'),
    'https://hyp-api.firebounce.today/v2/player?uuid=abc');
});

await t('path 和 query 原样保留（反代是透明转发）', () => {
  const u = BSK.rewriteUrl('https://api.hypixel.net/v2/skyblock/auctions?page=3&x=1');
  assert.ok(u.includes('/v2/skyblock/auctions'));
  assert.ok(u.includes('page=3'));
  assert.ok(u.includes('x=1'));
});

await t('非 Hypixel 的请求一动不动', () => {
  for (const u of ['https://example.com/a',
                   'https://api.mojang.com/users/profiles/minecraft/x',
                   'https://api.firebounce.today/api/denick?nick=x']) {
    assert.strictEqual(BSK.rewriteUrl(u), u, u + ' 不该被改');
  }
});

await t('★ 长得像的钓鱼域名不能被当成目标', () => {
  // 最容易写错的地方: 用 includes / endsWith 判断就会中招
  for (const u of ['https://api.hypixel.net.evil.com/v2/x',
                   'https://evilapi.hypixel.net.attacker.io/v2/x',
                   'https://notapi.hypixel.net/v2/x']) {
    assert.strictEqual(BSK.rewriteUrl(u), u, u + ' 不该被改');
  }
});

await t('已经是反代的地址不会被二次改写', () => {
  const u = 'https://hyp-api.firebounce.today/v2/player?uuid=abc';
  assert.strictEqual(BSK.rewriteUrl(u), u);
});

console.log('\nKey 策略');

/**
 * 取"注入进去的那个 key 值"。
 *
 * ★ 不能写死 header 名: 浏览器里为了过 CORS 预检用的是 X-API-Key
 *   (反代的白名单是 Authorization, X-API-Key, Content-Type),
 *   而 Node 没有 CORS 限制, 用的是更贴近官方习惯的 API-Key。
 *   所以这里按**值**找, 跟实现选哪个名字无关。
 */
function injectedKey(headers) {
  for (const name of ['X-API-Key', 'API-Key', 'Authorization']) {
    const v = headers.get ? headers.get(name) : (headers[name] || null);
    if (v) return v;
  }
  return null;
}

await t('没带 key 时注入我们的', () => {
  const r = BSK.buildFetchHeaders({ 'Content-Type': 'application/json' },
                                  'https://hyp-api.firebounce.today/v2/x');
  assert.strictEqual(r.injected, true);
  assert.strictEqual(injectedKey(r.headers), KEY);
});

await t('★ 强制覆盖: 调用方自己的 key 必须被删掉', () => {
  // 不删的话有些实现会优先读旧头, 等于没换
  const r = BSK.buildFetchHeaders({ 'API-Key': 'their-own-key' },
                                  'https://hyp-api.firebounce.today/v2/x');
  const values = [...r.headers.entries()].map((p) => p[1]);
  assert.ok(!values.includes('their-own-key'), '旧 key 还在: ' + JSON.stringify([...r.headers]));
  assert.strictEqual(injectedKey(r.headers), KEY);
});

await t('强制覆盖: Bearer 形式的 Authorization 也要删', () => {
  const r = BSK.buildFetchHeaders({ Authorization: 'Bearer their-own-key' },
                                  'https://hyp-api.firebounce.today/v2/x');
  const auth = r.headers.get('Authorization');
  assert.strictEqual(auth, null, 'Bearer 不该留着: ' + auth);
  assert.strictEqual(injectedKey(r.headers), KEY);
});

await t('不强制时: 调用方带了 key 就不动', () => {
  BSK.setConfig({ forceKey: false });
  const r = BSK.buildFetchHeaders({ 'API-Key': 'their-own-key' },
                                  'https://hyp-api.firebounce.today/v2/x');
  assert.strictEqual(r.injected, false);
  assert.strictEqual(r.hadKey, true);
  BSK.setConfig({ forceKey: true });
});

await t('不强制时: 调用方没带才补上', () => {
  BSK.setConfig({ forceKey: false });
  const r = BSK.buildFetchHeaders({}, 'https://hyp-api.firebounce.today/v2/x');
  assert.strictEqual(r.injected, true);
  BSK.setConfig({ forceKey: true });
});

await t('?key= 也算"调用方带了 key"', () => {
  assert.strictEqual(BSK.planKey([], 'https://hyp-api.firebounce.today/v2/x?key=zzz').hasKey, true);
});

await t('X-API-Key 形式也被认成"已带 key"', () => {
  assert.strictEqual(BSK.planKey([['X-API-Key', 'x']], 'https://h/v2/x').hasKey, true);
});

await t('没配 apiKey 时什么都不注入', () => {
  BSK.setConfig({ apiKey: '' });
  assert.strictEqual(BSK.buildFetchHeaders({}, 'https://hyp-api.firebounce.today/v2/x').injected, false);
  BSK.setConfig({ apiKey: KEY });
});

// ---------------------------------------------------------------------------
// 真发请求：起一个本地假反代，把 targetHosts 指到它，验证 fetch 补丁真的生效
// ---------------------------------------------------------------------------
console.log('\nfetch 拦截（真打本地假反代）');
{
  const http = require('node:http');
  const seen = [];
  const server = http.createServer((req, res) => {
    seen.push({ url: req.url, key: req.headers['x-api-key'] || req.headers['api-key'] || '' });
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ ok: true }));
  });
  await new Promise((r) => server.listen(0, '127.0.0.1', r));
  const port = server.address().port;
  const BASE = `http://127.0.0.1:${port}`;

  // 把"官方域名"当成 127.0.0.1，反代也指向它 —— 这样 fetch 补丁会真的介入
  BSK.install({
    proxyBase: BASE,
    targetHosts: ['127.0.0.1'],
    apiKey: 'bsk_LOCALTEST',
    forceKey: true,
  });

  await t('① fetch(url, init) 被拦且注入了 key', async () => {
    const r = await fetch(`${BASE}/v2/status`, { headers: { 'API-Key': 'their-key' } });
    await r.json();
    const last = seen[seen.length - 1];
    assert.strictEqual(last.url, '/v2/status');
    assert.strictEqual(last.key, 'bsk_LOCALTEST', 'key 应该是我们的');
  });

  await t('② fetch(Request) 也被拦且注入了 key', async () => {
    const req = new Request(`${BASE}/v2/player?uuid=x`, {
      headers: { 'API-Key': 'their-key' },
    });
    const r = await fetch(req);
    await r.json();
    const last = seen[seen.length - 1];
    assert.strictEqual(last.url, '/v2/player?uuid=x');
    assert.strictEqual(last.key, 'bsk_LOCALTEST');
  });

  await t('③ 非目标的请求不该被插手', async () => {
    const before = seen.length;
    const r = await fetch('https://example.com/');   // 会失败, 但重点是没被改写
    await r.text().catch(() => {});
    assert.strictEqual(seen.length, before, 'example.com 不该打到假反代');
  }).catch(() => {});

  await t('统计有记账', () => {
    const s = BSK.getStats();
    assert.ok(s.intercepted >= 2, 'intercepted=' + s.intercepted);
    assert.ok(s.keyInjected >= 2, 'keyInjected=' + s.keyInjected);
  });

  BSK.uninstall();
  await new Promise((r) => server.close(r));
}

console.log(`\n${pass} 通过, ${fail} 失败`);
process.exit(fail ? 1 : 0);

File: interceptor/tests/test_interceptor.py
# -*- coding: utf-8 -*-
"""
Python 拦截器测试。

    python -m unittest discover -s tests -v
    （或者直接 python tests/test_interceptor.py）

除了纯函数，最后一段会**真起一个本地假反代**，用 requests / urllib
真发一遍请求，确认补丁真的生效 —— 只测纯函数会漏掉"补丁压根没挂上"这种问题。
"""

from __future__ import annotations

import json
import os
import sys
import threading
import unittest
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import bsk_hypixel_interceptor as bsk  # noqa: E402

KEY = "bsk_TESTKEY"

try:
    import requests
    HAVE_REQUESTS = True
except ImportError:
    HAVE_REQUESTS = False


class UrlRewriteTest(unittest.TestCase):
    def setUp(self):
        bsk.uninstall()
        bsk.configure(proxy_base=bsk.PROXY_BASE, target_hosts=("api.hypixel.net",),
                      api_key=KEY, force_key=True)

    def test_official_host_is_rewritten(self):
        self.assertEqual(
            bsk.rewrite_url("https://api.hypixel.net/v2/player?uuid=abc"),
            "https://hyp-api.firebounce.today/v2/player?uuid=abc")

    def test_path_and_query_are_preserved(self):
        u = bsk.rewrite_url("https://api.hypixel.net/v2/skyblock/auctions?page=3&x=1")
        self.assertIn("/v2/skyblock/auctions", u)
        self.assertIn("page=3", u)
        self.assertIn("x=1", u)

    def test_other_hosts_untouched(self):
        for u in ("https://example.com/a",
                  "https://api.mojang.com/users/profiles/minecraft/x",
                  "https://api.firebounce.today/api/denick?nick=x"):
            self.assertEqual(bsk.rewrite_url(u), u, u)

    def test_lookalike_hosts_are_not_targets(self):
        """
        ★ 最容易写错的地方: 用 in / endswith 判断就会把钓鱼站当目标,
          等于主动把你的 bsk_ Key 送给攻击者。
        """
        for u in ("https://api.hypixel.net.evil.com/v2/x",
                  "https://evilapi.hypixel.net.attacker.io/v2/x",
                  "https://notapi.hypixel.net/v2/x"):
            self.assertEqual(bsk.rewrite_url(u), u, "%s 不该被改" % u)

    def test_proxy_url_not_rewritten_again(self):
        u = "https://hyp-api.firebounce.today/v2/player?uuid=abc"
        self.assertEqual(bsk.rewrite_url(u), u)


class KeyPolicyTest(unittest.TestCase):
    def setUp(self):
        bsk.uninstall()
        bsk.configure(proxy_base=bsk.PROXY_BASE, target_hosts=("api.hypixel.net",),
                      api_key=KEY, force_key=True)

    def test_injects_when_absent(self):
        h = bsk.apply_key({"Content-Type": "application/json"})
        self.assertEqual(h.get("API-Key"), KEY)

    def test_force_overwrites_and_removes_old_key(self):
        h = bsk.apply_key({"API-Key": "their-own-key"})
        self.assertEqual(h.get("API-Key"), KEY)
        self.assertNotIn("their-own-key", list(h.values()))

    def test_force_removes_bearer_authorization(self):
        h = bsk.apply_key({"Authorization": "Bearer their-own-key"})
        self.assertNotIn("Authorization", h)
        self.assertEqual(h.get("API-Key"), KEY)

    def test_x_api_key_also_removed(self):
        h = bsk.apply_key({"X-API-Key": "their-own-key"})
        self.assertNotIn("X-API-Key", h)
        self.assertEqual(h.get("API-Key"), KEY)

    def test_no_force_keeps_callers_key(self):
        bsk.configure(force_key=False)
        h = bsk.apply_key({"API-Key": "their-own-key"})
        self.assertEqual(h.get("API-Key"), "their-own-key")

    def test_no_force_fills_in_when_absent(self):
        bsk.configure(force_key=False)
        self.assertEqual(bsk.apply_key({}).get("API-Key"), KEY)

    def test_key_in_query_counts_as_present(self):
        bsk.configure(force_key=False)
        h = bsk.apply_key({}, "https://hyp-api.firebounce.today/v2/x?key=zzz")
        self.assertNotIn("API-Key", h)

    def test_empty_key_injects_nothing(self):
        bsk.configure(api_key="")
        self.assertNotIn("API-Key", bsk.apply_key({}))

    def test_headers_input_is_not_mutated(self):
        """别把调用方传进来的 dict 就地改了 —— 那是很难查的副作用。"""
        original = {"API-Key": "their-own-key"}
        bsk.apply_key(original)
        self.assertEqual(original, {"API-Key": "their-own-key"})

    def test_case_insensitive_header_names(self):
        h = bsk.apply_key({"aPi-kEy": "their-own-key"})
        self.assertEqual(h.get("API-Key"), KEY)
        self.assertNotIn("their-own-key", list(h.values()))


class ConfigTest(unittest.TestCase):
    def setUp(self):
        bsk.uninstall()
        bsk.configure(**bsk.DEFAULTS)

    def test_unknown_option_is_rejected(self):
        """打错字要报错, 不能静默忽略 —— 否则用户以为配上了。"""
        with self.assertRaises(KeyError):
            bsk.configure(api_kye="bsk_x")

    def test_stats_exposes_installed_state(self):
        s = bsk.stats()
        self.assertIn("installed", s)
        self.assertIn("intercepted", s)

    def test_mask_key(self):
        self.assertEqual(bsk.mask_key("short"), "short")
        self.assertIn("…", bsk.mask_key("bsk_" + "a" * 32))


# ---------------------------------------------------------------------------
# 真发请求：本地假反代
# ---------------------------------------------------------------------------

class _FakeProxy(BaseHTTPRequestHandler):
    seen = []

    def do_GET(self):
        _FakeProxy.seen.append({
            "path": self.path,
            "key": self.headers.get("API-Key") or self.headers.get("X-API-Key") or "",
            "auth": self.headers.get("Authorization") or "",
        })
        body = json.dumps({"ok": True, "path": self.path}).encode()
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, *a):
        pass


class LivePatchTest(unittest.TestCase):
    """
    ★ 真起一个假反代、真发请求 —— 只测纯函数会漏掉"补丁没挂上"。
    """

    @classmethod
    def setUpClass(cls):
        _FakeProxy.seen = []
        cls.server = ThreadingHTTPServer(("127.0.0.1", 0), _FakeProxy)
        cls.port = cls.server.server_address[1]
        cls.thread = threading.Thread(target=cls.server.serve_forever, daemon=True)
        cls.thread.start()
        cls.base = "http://127.0.0.1:%d" % cls.port

    @classmethod
    def tearDownClass(cls):
        cls.server.shutdown()

    def setUp(self):
        bsk.uninstall()
        _FakeProxy.seen = []
        # 把"官方域名"当成 127.0.0.1，反代也指向它 —— 这样补丁会真的介入
        bsk.install(proxy_base=self.base, target_hosts=("127.0.0.1",),
                    api_key="bsk_LOCALTEST", force_key=True)

    def tearDown(self):
        bsk.uninstall()

    @unittest.skipUnless(HAVE_REQUESTS, "没装 requests")
    def test_requests_is_intercepted_and_key_replaced(self):
        r = requests.get(self.base + "/v2/status", headers={"API-Key": "their-key"}, timeout=10)
        self.assertEqual(r.status_code, 200)
        self.assertEqual(len(_FakeProxy.seen), 1)
        self.assertEqual(_FakeProxy.seen[0]["key"], "bsk_LOCALTEST")
        self.assertEqual(_FakeProxy.seen[0]["path"], "/v2/status")

    @unittest.skipUnless(HAVE_REQUESTS, "没装 requests")
    def test_requests_with_bearer_is_replaced(self):
        requests.get(self.base + "/v2/player?uuid=x",
                     headers={"Authorization": "Bearer their-key"}, timeout=10)
        got = _FakeProxy.seen[-1]
        self.assertEqual(got["key"], "bsk_LOCALTEST")
        self.assertEqual(got["auth"], "", "旧的 Bearer 应该被删掉")

    @unittest.skipUnless(HAVE_REQUESTS, "没装 requests")
    def test_requests_session_also_intercepted(self):
        s = requests.Session()
        s.get(self.base + "/v2/x", timeout=10)
        self.assertEqual(_FakeProxy.seen[-1]["key"], "bsk_LOCALTEST")

    @unittest.skipUnless(HAVE_REQUESTS, "没装 requests")
    def test_non_target_is_not_intercepted(self):
        """非目标不该被塞 key。"""
        before = len(_FakeProxy.seen)
        try:
            requests.get("https://example.com/", timeout=5)
        except Exception:
            pass
        self.assertEqual(len(_FakeProxy.seen), before)

    def test_urllib_is_intercepted(self):
        import urllib.request
        with urllib.request.urlopen(self.base + "/v2/status", timeout=10) as r:
            self.assertEqual(r.status, 200)
        self.assertEqual(_FakeProxy.seen[-1]["key"], "bsk_LOCALTEST")

    def test_urllib_request_object_is_intercepted(self):
        import urllib.request
        req = urllib.request.Request(self.base + "/v2/y", headers={"API-Key": "their-key"})
        with urllib.request.urlopen(req, timeout=10) as r:
            self.assertEqual(r.status, 200)
        self.assertEqual(_FakeProxy.seen[-1]["key"], "bsk_LOCALTEST")

    @unittest.skipUnless(HAVE_REQUESTS, "没装 requests")
    def test_uninstall_restores(self):
        bsk.uninstall()
        _FakeProxy.seen = []
        # 还原之后, 调用方自己的 key 原样发出去
        requests.get(self.base + "/v2/z", headers={"API-Key": "their-key"}, timeout=10)
        self.assertEqual(_FakeProxy.seen[-1]["key"], "their-key")

    @unittest.skipUnless(HAVE_REQUESTS, "没装 requests")
    def test_context_manager_restores_config(self):
        bsk.uninstall()
        bsk.configure(api_key="bsk_OUTER")
        with bsk.intercept(api_key="bsk_INNER"):
            requests.get(self.base + "/v2/c", timeout=10)
            self.assertEqual(_FakeProxy.seen[-1]["key"], "bsk_INNER")
        self.assertEqual(bsk.stats()["config"]["api_key"], "bsk_OUTER")


if __name__ == "__main__":
    unittest.main(verbosity=2)
